### PR TITLE
gh-120417: Add more #noqa to used imports in the stdlib

### PR DESCRIPTION
**Pull request by vstinner**
Tools such as ruff can ignore "imported but unused" warnings if a line ends with "# noqa: F401". It avoids the temptation to remove an import which is used effectively.

* Issue: gh-120417

### CODE REVIEW

**Comment by vstinner** [comment]
cc @AlexWaygood @gpshead @serhiy-storchaka

**Comment by AlexWaygood** [comment]
I'm still against this approach, except for maybe `Lib/collections/abc.py`, where there are only two being `noqa`'d. If we need to ignore this many imports in a file, there are lots of other workarounds:
- Add `__all__` to the file so it's clear to the linter what's being deliberately re-exported
- Ignore the whole file by putting `# noqa: F401` on a line by itself near the top of the file
- Use a `--exclude` CLI flag when running the linter to exclude the whole file from the check.

In my opinion, this is just too ugly for a file like `ctypes/__init__.py`.

**Comment by ericsnowcurrently** [comment]
tl;dr I strongly favor using `__all__` for nearly all of these cases.

In a very few of the cases I'd say the linters aren't able to do a good job of identifying false positives for `F401`.  However, IMHO the linter is actually doing the right thing in the remaining cases.  In that vast majority it is simply exposing the fact that we aren't using `__all__` when we should be.  If we used `__all__` the way we should then there would be no need for `#noqa`, etc. in those cases.

Here we're dealing with a whole lot of bound objects that are meant to be available to users as part of a module's API.  `__all__` is exactly *the* place where you indicate which bound objects are available to users. [^1]  All objects that are imported just to be exposed as-is (I usually call them "aliases") fall into this category.  `__all__` is the natural place to identify them, to effectively say "these objects were imported to be exposed by this module."

[^1]: I'd argue that having `__all__` is a good idea generally; it makes the module's public API explicit to users/maintainers *and* to the runtime, available *on* the module for code to easily look at.

So why are we avoiding `__all__` for all these "alias" cases?

**Comment by serhiy-storchaka** [comment]
This is pretty common idiom -- defining the names in the submodules and exposing them as a public API at the package level.

Ideally, linters should be aware of this idiom.

**Comment by vstinner** [comment]
Here is a different approach to fix these warnings: https://github.com/python/cpython/pull/120461 PR adding ``__all__`` to modules. I left ctypes and ssl aside for now.

**Comment by vstinner** [comment]
I closed this PR: https://github.com/python/cpython/pull/120461 looks like a better approach.
